repair/repair.cc

```diff
@@ -1600,3 +1600,92 @@ future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_me
+        bool is_decommission = myip == leaving_node;
+        auto op = is_decommission ? "decommission_with_repair" : "removenode_with_repair";
+        rlogger.info("{}: started with keyspaces={}, leaving_node={}", op, keyspaces, leaving_node);
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                rlogger.info("{}: keyspace={} does not exist any more, ignoring it", op, keyspace_name);
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            // First get all ranges the leaving node is responsible for
+            dht::token_range_vector ranges = strat.get_ranges(leaving_node);
+            rlogger.info("{}: started with keyspace={}, leaving_node={}, nr_ranges={}", op, keyspace_name, leaving_node, ranges.size());
+            std::unordered_map<dht::token_range, std::vector<inet_address>> current_replica_endpoints;
+            // Find (for each range) all nodes that store replicas for these ranges as well
+            for (auto& r : ranges) {
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                auto eps = strat.calculate_natural_endpoints(end_token, tm);
+                current_replica_endpoints.emplace(r, std::move(eps));
+            }
+            auto temp = tm.clone_after_all_left();
+            // leaving_node might or might not be 'leaving'. If it was not leaving (that is, removenode
+            // command was used), it is still present in temp and must be removed.
+            if (temp.is_member(leaving_node)) {
+                temp.remove_endpoint(leaving_node);
+            }
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+            for (auto it = ranges.begin(); it != ranges.end();) {
+                auto& r = *it;
+                if (need_preempt()) {
+                    seastar::thread::yield();
+                }
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                const std::vector<inet_address> new_eps = ks.get_replication_strategy().calculate_natural_endpoints(end_token, temp);
+                const std::vector<inet_address>& current_eps = current_replica_endpoints[r];
+                std::unordered_set<inet_address> neighbors_set(new_eps.begin(), new_eps.end());
+                bool skip_this_range = false;
+                // For decommission operation, the decommission node will
+                // repair all the ranges the leaving node is responsible for.
+                if (!is_decommission) {
+                    // For removenode operation, the new owner of the range
+                    // will repair the data. Nodes that are not the new owner
+                    // of a range will ignore the range. There can be only one
+                    // new owner for each of the ranges.
```

Comment:
Wouldn't the decommission and leaving node be the same in the decommission case? 
Also, why do you have to complicate the removenode case? Why can't the node that received the removenode command repair all ranges belonging to the leaving node? 
Is it because of repair code can only repair ranges that the current node is a replica for?

Reply:
> Wouldn't the decommission and leaving node be the same in the decommission case?

Yes. 

> Also, why do you have to complicate the removenode case? Why can't the node that received the removenode command repair all ranges belonging to the leaving node?
> Is it because of repair code can only repair ranges that the current node is a replica for

Because this will cause more nodes to repair the same range. For example, for a range, old replicas: n1,n2,n3,after the removenode ops, new replicas: n1,n2,n4. With you method, both n1, n2, n4 will all repair the same range.

Reply:
> Then, please, fix the comment because a reader may mistakenly assume that there is a chance that the "decommissioned", the "leaving", and the node where this code runs may be different nodes.

The node runs the node can be a different node than leaving_node node, e.g., when you run removenode on n1 to remove the dead node n2. leaving_node will be n2. 

When you run decommission on n1, to remove n1 itself, leaving_node = n1.

What is confusing?

Reply:
> I thought that this code runs on the node where "nodetool removenode ..." was issued.
> 
> There are 2 things to do when you decommission/remove node:
> 
>     1. Stream data to new replicas.
> 
>     2. update all nodes in the cluster about the node that has left.
> 
> 
> It's going to be a lot easier if both are orchestrated from the same node.
> And it makes sense to choose the node where the corresponding nodetool command has been issued as such a node.

storage_service::restore_replica_count -> removenode_with_repair runs on all the nodes that owns the data after the node is removed by nodetool removenode. 

node that takes the nodetool decommission or removenode command is already the  coordinator that coordinates the removal of the node in question.

Reply:
> > Then, please, fix the comment because a reader may mistakenly assume that there is a chance that the "decommissioned", the "leaving", and the node where this code runs may be different nodes.
> 
> The node runs the node can be a different node than leaving_node node, e.g., when you run removenode on n1 to remove the dead node n2. leaving_node will be n2.
> 
> When you run decommission on n1, to remove n1 itself, leaving_node = n1.
> 
> What is confusing?

The comment that refers only to a decommissioning case where leaving==n1 always.

Reply:
> > I thought that this code runs on the node where "nodetool removenode ..." was issued.
> > There are 2 things to do when you decommission/remove node:
> > ```
> > 1. Stream data to new replicas.
> > 
> > 2. update all nodes in the cluster about the node that has left.
> > ```
> > 
> > 
> > It's going to be a lot easier if both are orchestrated from the same node.
> > And it makes sense to choose the node where the corresponding nodetool command has been issued as such a node.
> 
> storage_service::restore_replica_count -> removenode_with_repair runs on all the nodes that owns the data after the node is removed by nodetool removenode.
> 
> node that takes the nodetool decommission or removenode command is already the coordinator that coordinates the removal of the node in question.

Then why do we need this complexity? Why not to issue repairs from the coordinator? It's supposed to fix all relevant data on all relevant nodes, isn't it?

Reply:
No. For example, assume n1,n2,n3,n4,n5 in the cluster. n5 is dead and removed. n4 is the node runs nodetool removnode. Before a range is owned by n1,n2,n5. After it is owned by n1,n2,n3. The coordinator n4 does not own the range. Running repair n4 is not enough. 

Also, it will be more efficient if we can run repair in multiple nodes in parallel for the removenode operation.

Reply:
> > > Then, please, fix the comment because a reader may mistakenly assume that there is a chance that the "decommissioned", the "leaving", and the node where this code runs may be different nodes.
> > 
> > 
> > The node runs the node can be a different node than leaving_node node, e.g., when you run removenode on n1 to remove the dead node n2. leaving_node will be n2.
> > When you run decommission on n1, to remove n1 itself, leaving_node = n1.
> > What is confusing?
> 
> The comment that refers only to a decommissioning case where leaving==n1 always.

For decommission, the leaving node is always n1 (the one runs the nodetool decommission). What is wrong?